tools/dynamic_framework_slicer/dynamic_framework_slicer.sh

```diff
@@ -0,0 +1,80 @@
+#!/bin/bash 
+set -eu
+
+declare -a BINARIES
+IN=""
+OUT=""
+while [[ $# -gt 0 ]]; do
+  arg="$1"
+  case $arg in
+      --in)
+      readonly IN="$2"
+      shift
+      ;;
+      --out)
+      readonly OUT="$2"
+      shift
+      ;;
+      *)
+      BINARIES+=("$1")
+      ;;
+  esac
+  shift
+done
+
+if [[ "$IN" == "" || "$OUT" == "" || "$BINARIES" == "" ]]; then
+    echo "Usage: $0 --in path/to/infile --out path/to/outfile binaries..."
+    exit 1
+fi
+
+if [[ ! -f "$IN" ]]; then
+    echo "Expected regular file with --in"
+    exit 1
+fi
+
+# Strip out any unnecessary slices from embedded dynamic frameworks to save space
+
+# Gather all binary slices
+declare -a all_bin_slices
+for bin in "${BINARIES[@]}"; do
+    # Lets figure out what slices the app has
+    bin_slices=$(xcrun lipo -info "$bin" | cut -d: -f3)
+    
+    # Add the app slices to the overall list of required slices
+    for slice in $bin_slices ; do
+        # store the slice if we don't already have it
+        if [[ "${#all_bin_slices[@]}" -eq 0 || $( echo "${all_bin_slices[@]}" | grep -q "$slice") ]] ; then
+            all_bin_slices+=($slice)
+        fi
+    done
+done
+
+# Gather the slices in the framework
+declare -a framework_slices
+for slice in $(xcrun lipo -info "$IN" | cut -d: -f3); do
+    framework_slices+=($slice)
+done
+
+# Figure out what, if anything, we should strip
+declare -a slices_needed
+just_copy=1
+for slice in "${framework_slices[@]}"; do
+    if echo "${all_bin_slices[@]}" | grep -q "$slice" ; then
+        slices_needed+=($slice)
+    else
+        # We have a slice in the framework that does not exist in any binary
+        # so a plain copy wont work
+        just_copy=0
```

Comment:
If the `otool` method is used, and if `slices_needed` == `all_bin_slices` does hold, then I think this block (calculating `slices_needed` and `just_copy`), and the last block, can be merged and reduced to:

```bash
if [[ "${all_bin_slices[*]}" == "${framework_slices[*]}" ]]; then
    cp "$IN" "$OUT"
else
    declare -a lipo_args
    for slice in "${all_bin_slices[@]}"; do
        lipo_args+=(-extract $slice)
    done
    xcrun lipo "$IN" "${lipo_args[@]}" -output "$OUT"
fi
```

Reply:
It isn't quite that simple, as @thomasvl  points out above. It is possible that a binary may have a slice that doesn't exist in the current framework, so we need to figure out what the minimum set of slices, that exist in the current framework, that we want to extract.

I have changed the logic to use something similar to the suggested otool method.